### ISSUE TITLE
WaveNet implementation for image generation (based in this repository) 

### ISSUE THREAD

**Issue by Zeta36** [ibab/tensorflow-wavenet#129]
As I told in #125, I've made a revision of this project where a feed the model using the gray-scale channel of a dataset of images (similar **but no equal** to what DeepMind made with **PixelCNN**).

I modify the TextReader I made for the WaveNet text generation version (#117). Specifically: I made a ImageReader. So this reader now can load (using Pillow: PIL) image files, rescale them to for example 64x64 (because I have no GPU and no time to waste :P) (pic.resize((32,24), Image.ANTIALIAS)) and finally get the 8bit grayscale channel (using convert('L')).

I really love the theoretical idea after the generative model using dilated causal convolution layers!! The results are wonderful and easily extrapolated.

I made the next test with this image implementation: 

1) I used a small dataset with 12 similar (but different) images of a person (the woman in the picture is my mother). She is posing in different ways.
![20140211_202000](https://cloud.githubusercontent.com/assets/17341905/19213557/c6087b66-8d66-11e6-9ba1-93bd6bfd337c.jpg)

Above one of the original pictures, and below the scaled gray-scale image fed into the training time:
![example_train_file](https://cloud.githubusercontent.com/assets/17341905/19213573/09cfec26-8d67-11e6-9f35-22e49e62eccd.jpg)

2) I trained the model using a SAMPLE_SIZE = 4096 and a learning rate of 0.001

3) The loss dropped down gradually and only after 4k steps it was around ~0.09. I trained once again with learning rate = 0.0001, and after another 4k steps loss was around ~0.01

4) I generated the images using always a WINDOW = 4096 and a always 4096 samples (64x64 pixels so I can save a full image).

Well:

5) If a let the first sample as random: 
`waveform = np.random.randint(quantization_channels, size=(1,)).tolist()`
or if I let the image grow using probability distribution and not argmax:
`sample = np.random.choice(np.arange(quantization_channels), p=prediction)`

I juts get ugly noise:
![mam3000_06](https://cloud.githubusercontent.com/assets/17341905/19213581/52df12d4-8d67-11e6-8225-d01399b3f717.gif)
![mam3500_02](https://cloud.githubusercontent.com/assets/17341905/19213582/591745e0-8d67-11e6-8516-f4bd05bec6d4.gif)

But, and this is interesting:

6) If I set the first sample to the first pixel in each image from the dataset. For example:
`waveform = [169]`
and if I use argmax to select every sample:
`sample = np.argmax(prediction)`

I can generate each and every of the posing images from the dataset!!
![mam9000_0 01](https://cloud.githubusercontent.com/assets/17341905/19213605/b3dad064-8d67-11e6-959e-4561db9f56fb.jpg)
![mam9000_v2_0 01](https://cloud.githubusercontent.com/assets/17341905/19213607/b3dece62-8d67-11e6-80fb-624b0d3b280b.gif)
![mam9000_v3](https://cloud.githubusercontent.com/assets/17341905/19213609/b3df5d96-8d67-11e6-9628-94f4e1e91b52.gif)
![mam9000_v4_](https://cloud.githubusercontent.com/assets/17341905/19213606/b3dd69dc-8d67-11e6-829b-ee9e2df00f82.gif)
![mam9000_v5](https://cloud.githubusercontent.com/assets/17341905/19213608/b3ded74a-8d67-11e6-9293-5e32a131358e.gif)
![mam9000_v6](https://cloud.githubusercontent.com/assets/17341905/19213610/b3df491e-8d67-11e6-80fd-af802a8c1c98.gif)
![mam9000_v7](https://cloud.githubusercontent.com/assets/17341905/19213611/b40ef10a-8d67-11e6-8fc2-8db14ea8c10e.gif)

The model is wonderful, and it able to memorize a lot of information in its nodes!! I'm pretty sure something similar is what happens in our memory (in the brain). When we see an image, probably a "training" process will take place in some mini-neural network in our brain, and later when we need to remember that image (or text, or sound,...), another sub-network of our brain must to generate the memorized information and pass the data to other subnet of our brain.

Well. I hope this can help in anyway to the project we are trying to build in here.

I think I will try to implement the global conditioned WaveNet model and try to be able to select the image a want to recover just by passing the ID from the dataset image.

If somebody want to play with this, here you can get the code: https://github.com/Zeta36/tensorflow-image-wavenet

Regards,
Samu.

**Comment by nakosung** [comment]
Does it seem to be overfitted? 12 x 4096 only requires 64K floats and wavenet has much more parameters to fit in. If you can have similar result with hundreds of photos, it would seem to be meaningful.
